onnx/defs/shape_inference.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "data_type_utils.h"
+
+namespace ONNX_NAMESPACE {
+
+struct InferenceContext {
+  virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
+  virtual size_t getNumInputTypes() const = 0;
+  virtual const TypeProto_Tensor* getInputType(size_t index) const = 0;
+  virtual size_t getNumOutputTypes() const = 0;
+  virtual TypeProto_Tensor* getOutputType(size_t index) = 0;
+  virtual ~InferenceContext() {}
+};
+
+typedef void (*InferenceFunction)(InferenceContext&);
+
+template<typename T> inline google::protobuf::RepeatedField<T> retrieveValues(const AttributeProto* attr);
+template<> inline google::protobuf::RepeatedField<int64_t> retrieveValues(const AttributeProto* attr) { return attr->ints(); }
+
+template<typename T>
+inline bool getRepeatedAttribute(InferenceContext& ctx,
+                                 std::string attr_name,
+                                 std::vector<T>& values) {
+  const auto* attr = ctx.getAttribute(attr_name);
+  if (attr) {
+    for (const auto& value : retrieveValues<T>(attr)) {
+      values.push_back(value);
+    }
+    return true;
+  } else {
+    return false;
+  }
+
+}
+
+inline void propagateElemTypeFromInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex) {
```

Comment:
same?

Reply:
Why not putting these into the InferenceContext class? They are tightly tied to InferenceContext and there is no way to use them.
@linkerzhang

Reply:
I think either way is acceptable. @linkerzhang has asked that we keep them out of InferenceContext though (I guess so the API exposed to 3rd parties is smaller), so I am keeping it separate